### PR TITLE
Add deleting interview client part

### PR DESCRIPTION
**Pull request by dbuksha**
PR with params for mmb: https://github.com/tolstoyevsky/mmb/pull/123

### CODE REVIEW

**Comment by dbuksha-wis** [comment]
https://github.com/tolstoyevsky/mmb/pull/123

File: README.md [hunk 1]
```diff
@@ -51,6 +51,11 @@ Then add **hubot-huntflow-reloaded** to your `external-scripts.json`:
 |`REDIS_PORT`                | Port Redis listens on.                                                                        | 16379                     |
 |`REDIS_PASSWORD`            | Specifies the Redis password.                                                                 | null                      |
 |`REDIS_CHANNEL`             | Defines the name of Redis channel to get messages from the server.                            | `hubot-huntflow-reloaded` |
+|`BASE_SERVER_URL`           | Defines the server url to handle requests.                                                    | `http://127.0.0.1:8888/`  |
+|`SERVER_USER_EMAIL`         | Defines the server user email to make authorized requests.                                    | null                      |
+|`SERVER_USER_PASSWORD`      | Defines the server user password to make authorized requests.                                 | null                      |
+
+**Note:**  Be sure that  you specified `SERVER_USER_EMAIL` and `SERVER_USER_PASSWORD` to handle huntflow control requests.
```
Comment (polina-popova) [inline README.md]: Let's include the warning in the 'Prerequisites' section above, where other required params are listed.

File: client/axios-instance.js [hunk 2]
```diff
@@ -0,0 +1,114 @@
...
+  baseURL: BASE_SERVER_URL,
+  timeout: 1000
+})
+
+const isTokenAlive = (token) => {
+  if (!token) return false
+
+  let decoded = jwtDecode(token)
+  return decoded.exp > (new Date().getTime() / 1000)
+}
```
Comment (polina-popova) [inline client/axios-instance.js:10]: Do we really need to decode the token on the client side? First of all, I am not sure that it will work with real (not mocked) secret for the tokens' encoding, because you don't know it on the client side. Also now you can distinguish if the access token is expired by the error code.

File: client/interview-controller.js [hunk 3]
```diff
@@ -0,0 +1,55 @@
...
+//    end admin
+//  end group
+//
+
+const routines = require('hubot-routines')
+const utils = require('./utils')
+
+const interviewsInstance = require('./interviews-service')
+
+const candidateRegExp = new RegExp(/([А-Я][а-я]+)\s([А-Я][а-я]+)/g)
```
Comment (polina-popova) [inline client/interview-controller.js]: The script will be used with other Hubot scripts so it would be better to make more specific regexp, to avoid mixes (for example `удалить интервью кандидата name`).

File: client/interview-controller.js [hunk 4]
```diff
@@ -0,0 +1,55 @@
...
+      .then(function (data) {
+        candidates = data
+        if (candidates.length === 0) return msg.send(utils.MSG_NOT_ANY_SCHEDULED_INTERVIEWS)
+        return msg.send(buildCandidatesButtons(data))
+      })
+      .catch((error) => msg.send(error))
+  })
+
+  robot.respond(candidateRegExp, async (msg) => {
+    const cand = candidates.find(user => msg.match[0].match(/[А-Яа-я]+/g)[0] === user.last_name)
```
Comment (polina-popova) [inline client/interview-controller.js]: What if we have two candidates with the same surnames?

File: client/axios-instance.js [hunk 5]
```diff
@@ -0,0 +1,114 @@
...
+  }
+})
+
+// response
+instance.interceptors.response.use((response) => {
+  return response
+}, (error) => {
+  const originalRequest = error.config
+
+  if (originalRequest._retry) return Promise.reject(error)
```
Comment (polina-popova) [inline client/axios-instance.js:70]: You can use the `rave` function from utils to exclude the error message from the info log stream.

**Review by polina-popova — CHANGES_REQUESTED** [review]

File: client/axios-instance.js [hunk 6]
```diff
@@ -0,0 +1,114 @@
...
+  return decoded.exp > (new Date().getTime() / 1000)
+}
+
+const setTokens = (tokens) => {
+  accessToken = tokens.access
+  refreshToken = tokens.refresh
+}
+
+const getAccessToken = () => {
+  return new Promise((resolve, reject) => {
```
Comment (ya7on) [inline client/axios-instance.js]: Why are you using promises here? You can already use axios with `async/await` contruction or with `then/catch`

File: client/interviews-service.js [hunk 7]
```diff
@@ -0,0 +1,17 @@
+const instance = require('./axios-instance')
+
+class InterviewsService {
```
Comment (ya7on) [inline client/interviews-service.js]: Class declaration is not needed here. You even not using context. This methods are static

File: index.js [hunk 8]
```diff
@@ -5,4 +5,5 @@ const path = require('path')
 module.exports = (robot) => {
   const scriptsPath = path.resolve(__dirname, 'client')
   robot.loadFile(scriptsPath, 'huntflow-reloaded.js')
+  robot.loadFile(scriptsPath, 'interview-controller.js')
```
Comment (ya7on) [inline index.js:8]: Why do we need two entry points to the application?

File: client/axios-instance.js [hunk 9]
```diff
@@ -0,0 +1,114 @@
...
+// init axios instance with base url
+const instance = axios.create({
+  baseURL: BASE_SERVER_URL,
+  timeout: 1000
+})
+
+const isTokenAlive = (token) => {
+  if (!token) return false
+
+  let decoded = jwtDecode(token)
```
Comment (ya7on) [inline client/axios-instance.js]: Do not use `let` with values ​​that do not change. You have `const`

File: client/utils.js [hunk 10]
```diff
@@ -1,10 +1,28 @@
 const moment = require('moment')
 
+// init exports instance
+const exp = module.exports = {}
+
+// env consts
+exp.BASE_SERVER_URL = process.env.BASE_SERVER_URL || 'http://127.0.0.1:8888'
+exp.SERVER_USER_EMAIL = process.env.SERVER_USER_PASSWORD
```
Comment (polina-popova) [inline client/utils.js]: It should be `process.env.SERVER_USER_EMAIL`

File: client/interview-controller.js [hunk 11]
```diff
@@ -0,0 +1,79 @@
...
+    routines.rave(robot, utils.MSG_AUTHORIZATION_DATA_NEEDED)
+    return
+  }
+
+  try {
+    await interviewsService.getAccessToken()
+  } catch (error) {
+    routines.rave(robot, utils.MSG_AUTHORIZATION_FAILED)
+    return
+  }
```
Comment (polina-popova) [inline client/interview-controller.js]: Our server is dependent on Redis, but Redis is run by the client. If we place retrieving tokens here we create some kind of race situation. Let's trigger the token receiving by the first command call.

File: client/interview-controller.js [hunk 12]
```diff
@@ -0,0 +1,79 @@
+// Description:
+//   A Hubot script which helps hrs to control interviews.
+//
+// Commands:
+//  begin group Huntflow Reloaded
+//    begin admin
+//      hubot удалить интервью - remove an interview from the list
+//    end admin
+//  end group
+//
```
Comment (polina-popova) [inline client/interview-controller.js:10]: For this configuration, the hubot-help script is not ready. So it fails with an error. Probably, for now, would be better to remove it.

File: client/interview-controller.js [hunk 13]
```diff
@@ -0,0 +1,79 @@
...
+    return msg.send(message)
+  })
+
+  robot.respond(deleteCandidateRegExp, async (msg) => {
+    let message = utils.MSG_ERROR_TRY_AGAIN
+
+    try {
+      const text = msg.match[0]
+      const candidate = text.split(' ').splice(4) // get elements after word 'кандидата'
+      await interviewsService.deleteCandidateInterview({ first_name: candidate[0], last_name: candidate[1] })
```
Comment (polina-popova) [inline client/interview-controller.js]: I can't delete interviews because there is a mix of names order: `candidate[0]` is the last name.

**Review by polina-popova — CHANGES_REQUESTED** [review]
Also to the `token/refresh` endpoint come request with `refresh=underfined` and it doesn't work.

File: client/interview-controller.js (hunk 12 above)
Comment (polina-popova) [reply client/interview-controller.js:10]: The PR to the hubot-help script is ready so you can keep it here. 
And let's rephrase it to `removes the non-expired interview of the specified candidate`

**Review by ya7on — CHANGES_REQUESTED** [review]
I have an error `Cannot read property 'status' of undefined` on message `удалить интервью кандидата ***`.

File: README.md [hunk 14]
```diff
@@ -23,6 +23,7 @@ The project is split into two parts: server and client. The server side is a [Hu
 
 * The bot must be in the channel specified via the `HUNTFLOW_REMINDER_CHANNEL` environment variable.
 * There must be specified the `POSTGRES_PASSWORD` environment variable (see the [configuration section](#configuration-1) for details).
+* There must be specified `SERVER_USER_EMAIL` and `SERVER_USER_PASSWORD`.
```
Comment (polina-popova) [inline README.md]: Let's add here the link to the relevant configuration section as above (note that we have two configuration sections, don't mix them).

File: client/interview-controller.js [hunk 15]
```diff
@@ -0,0 +1,72 @@
+// Description:
+//   A Hubot script which helps hrs to control interviews.
+//
+// Commands:
+//  begin group Huntflow Reloaded
+//    begin admin
+//      hubot удалить интервью - remove an interview from the list
```
Comment (polina-popova) [inline client/interview-controller.js]: Let's replace the description with `removes the non-expired interview of the specified candidate`.

File: client/interview-controller.js [hunk 16]
```diff
@@ -0,0 +1,72 @@
...
+    return
+  }
+
+  robot.respond(/(удалить интервью)\s*$/i, async (msg) => {
+    if (!await routines.isAdmin(robot, msg.message.user.name.toString())) {
+      msg.send(utils.MSG_PERMISSION_DENIED)
+      return
+    }
+
+    let message = utils.MSG_ERROR_TRY_AGAIN
```
Comment (polina-popova) [inline client/interview-controller.js:32]: Let's use it only if retry can help. Otherwise would be reasonable just retrieve the new token pair.

File: client/utils.js [hunk 17]
```diff
@@ -1,10 +1,28 @@
...
+// init exports instance
+const exp = module.exports = {}
+
+// env consts
+exp.BASE_SERVER_URL = process.env.BASE_SERVER_URL || 'http://127.0.0.1:8888'
+exp.SERVER_USER_EMAIL = process.env.SERVER_USER_EMAIL
+exp.SERVER_USER_PASSWORD = process.env.SERVER_USER_PASSWORD
+
+// messages constants
+exp.MSG_PERMISSION_DENIED = 'Недостаточно прав.'
```
Comment (polina-popova) [inline client/utils.js]: Let's use the same error message as in other scripts: `У тебя недостаточно прав для этой команды :rolling_eyes:`.

**Review by polina-popova — CHANGES_REQUESTED** [review]

File: client/utils.js [hunk 18]
```diff
@@ -1,10 +1,34 @@
...
+// env consts
+exp.BASE_SERVER_URL = process.env.BASE_SERVER_URL || 'http://127.0.0.1:8888'
+exp.SERVER_USER_EMAIL = process.env.SERVER_USER_EMAIL
+exp.SERVER_USER_PASSWORD = process.env.SERVER_USER_PASSWORD
+
+// messages constants
+exp.MSG_PERMISSION_DENIED = 'У тебя недостаточно прав для этой команды :rolling_eyes:'
+exp.MSG_NOT_ANY_SCHEDULED_INTERVIEWS = 'Запланированных интервью нет.'
+exp.MSG_ERROR_TRY_AGAIN = 'Произошла ошибка, попробуйте еще раз.'
+exp.MSG_SUCCESSFULLY_DELETED = 'Успешно удалено'
```
Comment (polina-popova) [inline client/utils.js]: Let's add the dot in the end.

File: client/utils.js [hunk 19]
```diff
@@ -1,10 +1,34 @@
...
+exp.MSG_AUTHORIZATION_DATA_NEEDED = 'SERVER_USER_EMAIL and SERVER_USER_PASSWORD ' +
+  'are mandatory parameters, however they\'re not specified.'
+exp.MSG_CHOOSE_CANDIDATE = 'Выберите одного из кандидатов, чтобы удалить интервью:'
+
+exp.ERROR_MSGS_FROM_SERVER = {
+  'invalid_auth_creds': 'Авторизация не удалась, обратитесь к администратору Rocket.Chat.',
+  'no_candidate': 'Кандидат с такими данными не был найден. (Возможно, ' +
+    'вы ввели имя и фамилию не в той последовательности?)',
```
Comment (polina-popova) [inline client/utils.js]: I'd better remove the text in braces because users supposed to push buttons.

**Review by polina-popova — CHANGES_REQUESTED** [review]
I added some minor changes requests and let's clean up the commit message body.

**Review by ya7on — APPROVED** [review]

**Review by polina-popova — APPROVED** [review]

File: client/interview-controller.js [hunk 20]
```diff
@@ -0,0 +1,77 @@
...
+    }
+
+    return msg.send(message)
+  })
+}
+
+function buildCandidatesButtons (users) {
+  const usersArray = users.map(function (user) {
+    const username = `${user.last_name} ${user.first_name}`
+    return [username, `удалить интервью кандидата ${username}`]
```
Comment (polina-popova) [inline client/interview-controller.js]: Let's start it with a big letter like in proper sentences.
